keps/sig-multicluster/4322-cluster-inventory/README.md

```diff
@@ -0,0 +1,993 @@
+manager.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+We try to summarize the most common properties that should be able to
+represent a cluster and support the above use case at the minimum scope.
+
+The target consumers of the API are users who manage the clusters and
+other tools to understand the clusters concept from the API for
+multicluster scheduling, workload distribution and cluster management.
+The API aims to provide the information for the consumers to answer the
+below questions:
+
+* Is the cluster under management?
+* How can I select a cluster?
+* Is the cluster healthy?
+* Does the cluster have certain capabilities or properties?
+* Does the cluster have sufficient resources?
+
+### Cluster Name
+
+It is required that cluster name is unique for each cluster, and it
+should also be unique among different providers (cluster manager). There
+are two option to ensure it is uniqueness
+
+#### Option 1
+
+The `metadata.name` of the cluster should never be set upon creation.
+Only `metadata.generateName` can be set.
+
+#### Option 2
+
+The `metadata.name` and `metadata.generateName` must have prefix which
+should be the same as the `spec.clusterManager.name`. Different cluster
+manager must set a different value of `spec.clusterManager.name` when
+the cluster is created.
+
+These two options can be achieved by a validating webhook or a
+`ValidatingAdmissionPolicy` resource.
+
+### Spec
+
+#### Display name
+
+It is a human-readable name of the cluster set by the consumer of the
+cluster.
+
+#### Cluster Manager
+
+An immutable field set by a cluster manager when this cluster resources
+is created by it. Each cluster manager instance should set a different
+values to this field.
+
+In addition, a predefined label with key "x-k8s.io/cluster-manager" needs
+to be added by the cluster manager upon creation. The value of the label
+MUST be the same as the name of the cluster manager. The purpose of this
+label is to make filter clusters from different cluster managers eaiser.
+
+### Status
```

Comment:
I disagree with Status, it should be Spec.
as CRD"s usually have a Status field dedicated to event's on the object.
Look at https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#create-a-customresourcedefinition, where `version` is inside `spec` not Status

The status subresource of a CRD can be considered to have the same objective of non-custom resources. While the spec defines the desired state of that particular resource, basically I declare what I want, the status instead explains the current situation of the resource I declared on the cluster and should help in understanding what is different between the desired state and the actual state.

Like a StatefulSet spec could say I want 3 replicas and its status say that right now only 1 of those replicas is ready and the next one is still starting, a custom resource status may tell me what is the current situation of whatever I declared in the specs.

Reply:
Here the version field is kind of static metadata. It is not configurable, which is quite different from those in `spec`.

In this KEP, cluster inventory API is not used to provision a cluster from scratch. That is what `cluster-api` has already done.

Reply:
I am referring to the 'status' field. I know this API is not for provisioning, but for living clusters. Still I think 'status' should not be used as proposed here.

Reply:
I think the field currently defined in the status is actual state but not desired state. These fields in status is supposed to be updated by the cluster manager but not configured by the consumer.

Reply:
I think my initial comment is miss-understood, My point is that I think `status` field is not being used as the default way CRD's use status. Look at https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#create-a-customresourcedefinition

Reply:
@ArangoGutierrez  I still don't follow. As @qiujian16 and @dixudx explained, the clusterAPI status fields are not desired state of the cluster. Where do you think we should put them then?

Reply:
So currently this proposes:
```yaml
spec:
  displayName: some-cluster
  clusterManager:
    name: some-cluster-manager
status:
 version:
   kubernetes: 1.28.0
 resources: # total resources of a cluster
   capacity:
     cpu: 12
     memory: 12Gi
   allocatable:
     cpu: 10
     memory: 16Gi
```

everything in Status, should be in Spec.

And Status be just for 
```yaml
 conditions:
   - type: ControlPlaneHealthy
     status: True
     lastTransitionTime: "2023-05-08T07:56:55Z"
     message: ""
   - type: AllNodesHealthy
     status: True
     lastTransitionTime: "2023-05-08T07:57:55Z"
     message: ""
   - type: Joined
     status: True
     lastTransitionTime: "2023-05-08T07:58:55Z"
     message: ""
   ```

Reply:
Cluster resources mutate, more in cloud environments were nodes can be added/removed dynamically. In my use case, nodes are added and removed on a daily basis 
Same for Version, what if the cluster if Updated?

Reply:
that is my point, but this shouldn't be mutated by user, but the cluster manager.

Reply:
Why the cluster manager? resource information, and other features should be advertise at the cluster level.
Question(maybe is not clear to me yet): where do you intend this API to live, on each cluster or in the management cluster?

Reply:
It is supposed to be live in a management cluster.

Reply:
I see, then my view was wrong.  I will remove some one my comments then